### DISCUSSION TITLE
Rate limiter socket

### DISCUSSION THREAD

**Discussion by mathysth** [rhinobase/hono-rate-limiter#15] [Q&A] [answered]
Hey i'm using hono rate limiter with redis and it work perfectly for me except for socket.
So I was wondering does this library support socket rate limiter ?
I have attached my socket to the server with the following  but I can spam the route as I want without any restriction:
```ts
this.server.hono.get(
      '/ws',
      upgradeWebSocket(() => {
        return {
          onMessage: async (event, ws) => {
            const response = await this.triggerEvent(event.data);
            if (response || response === false) {
              ws.send(JSON.stringify(response));
            }
          },
        }
      })
    )
```

**Comment by MathurAditya724** [comment]
Can you share a reproduction of this? Which runtime are you using?

So, I believe what is happening here is that `hono-rate-limiter` limits the requests coming from a user through REST API. But in the case of `WebSockets`, a request occurs only once and the subsequent requests are happening through the `WebSocket` protocol, which currently `hono-rate-limiter` is not keeping track of. It limits the number of connections a single user can open in an interval, as that is happening through the REST API.

We can add a helper function, which will run inside the `onMessage` function or wrap around the `createEvent` function. So every time a request using the `WebSocket` protocol comes through, we can track it and limit it based on the value from the `keyGenerator`. we can call it `socketLimiter` 😂

What do you think? We will need to figure out the API but the implementation seems easy

**Reply by mathysth** [reply]
Hey,

I'm using Bun.
And for the reproduction you can just use this [section on hono doc](https://hono.dev/helpers/websocket).
I use it with a custom event handler.
About the implementation, I think this is a good one. I saw a [similar anwser](https://stackoverflow.com/a/61821686/15431338) with socket.io for express and this may be helpfull for other I think

**Reply by MathurAditya724** [reply]
Sounds good! I need to figure out the API, mainly how devs will interact with it as I want to use the same props.

**Reply by mathysth** [reply]
Thanks,
Just to know will it take weeks or month to implement to know if I make a workaround

**Reply by MathurAditya724** [reply]
If you can help me test it out, we can finish it in this week only

**Reply by mathysth** [reply]
Yes sure ping me when needed

**Reply by MathurAditya724** [reply]
Hey, you can check the progress here #17, do let me know if you have any more ideas 😄

**Reply by mathysth** [reply]
At the moment, I have no further idea about what we discussed earlier. However, if you need to test and look for bugs once you're done, don't hesitate to ping me.

**Reply by MathurAditya724** [reply]
Hey, I have just released a new version in the `RC` channel `hono-rate-limiter@0.4.0-rc.1`, you can check out the docs in the [README.md](https://github.com/rhinobase/hono-rate-limiter/blob/feat/websocket/README.md).

I have tested it, need to write tests tho. Are there any good testing libs like `supertest` but for WebSockets? I will be merging it after that

**Reply by mathysth** [reply]
I have never done testing with ws but after some research, I found 3 answers.
Either you can use the following package but it hasn't been updated since 2 years : 
[superwstest](https://www.npmjs.com/package/superwstest)
Or you can setup your test with ws package:
[ws](https://www.npmjs.com/package/ws)
Or you can follow this guide : 
[ws](https://thomason-isaiah.medium.com/writing-integration-tests-for-websocket-servers-using-jest-and-ws-8e5c61726b2a#9e4d)

What do you think?

**Reply by MathurAditya724** [reply]
Yeah, I will have to see, I am thinking of going with `superwstest` just because it is compatible with `supertest` but thanks for the suggestion.

Were you able to try the RC? Any errors or feedback?

**Reply by mathysth** [reply]
I was able to have the limiter working but didn't have the time to really use it and find some problem. I will do that today

**Reply by mathysth** [reply]
@MathurAditya724 I just implemented the socket limiter but in my opinion, when the client has exceeded the limit he gets disconnected, can we return a code or message so we can handle it from the front ?
Are we forced to disconnect the user can't we do the same as the basic limiter ?

**Reply by MathurAditya724** [reply]
Hey, the API for the basic rate limiter and the WebSocket one are the same. So, you can use the `handler` function to customize the behavior.

**Reply by mathysth** [reply]
Yes I also saw that you let us handle the behavior of the onClose so I will do a custom implementation, thanks you.
For me it's good was a pleasure to help you

**Reply by MathurAditya724** [reply, marked as answer]
Thank you! I will try to merge this soon (have to finish those tests). Do let me know if you have any more ideas.

**Reply by mathysth** [reply]
Yes I will create another discussion or issue if I have any idea in the future cause I plan on using it for my project

**Comment by alexgleason** [comment]
This looks good. I also want to limit the number of concurrent connections to the socket by a user. So for example user `a` cannot have more than 5 active connections to the socket at once.

**Reply by alexgleason** [reply]
I already decided this is wrong. Having a connection open is basically free, it's the messages that are the problem.

**Reply by alexgleason** [reply]
I have a problem where someone is opening 40 websocket connections on my server and then piping /dev/urandom into them, and this destroys my server. I am not completely sure how that translates to individual websocket messages. It seems like both the rate and the size would matter.

**Reply by MathurAditya724** [reply]
You can use the `rateLimiter` and `websocketLimiter` together to limit the number of connections from a user and messages/operations from the socket respectively.

**Reply by alexgleason** [reply]
Thanks, that's exactly what I plan to do.

I may need to do more than that to mitigate this particular attack. Apparently websocket frames can be 16MB+ in size? I need to test locally to see exactly what happens on the Deno side when you pipe megabytes of trash into the socket.

Thank you for this library. It was very easy to get it going! https://gitlab.com/soapbox-pub/ditto/-/merge_requests/390

**Reply by MathurAditya724** [reply]
Interesting! you also have a custom function in the `websocketLimiter` to limit the messages. If you check the size of the messages in that function other than the limiting stuff, you should be good to go!
